presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java

```diff
@@ -579,12 +579,17 @@ public static long varcharToShortDecimal(Slice value, long precision, long scale
     @UsedByGeneratedCode
     public static Slice varcharToLongDecimal(Slice value, long precision, long scale, BigInteger tenToScale)
     {
-        String stringValue = value.toString(UTF_8);
-        BigDecimal decimal = new BigDecimal(stringValue).setScale(intScale(scale), HALF_UP);
-        if (overflows(decimal, precision)) {
-            throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast VARCHAR '%s' to DECIMAL(%s, %s)", stringValue, precision, scale));
+        try {
```

Comment:
I'd restructure as:

```java
BigDecimal result;
String stringValue = value.toString(UTF_8);
boolean failed = false;
try {
    result = new BigDecimal(stringValue).setScale(intScale(scale), HALF_UP);
}
catch (NumberFormatException e) {
    failed = true;
}

if (failed || overflows(result, precision)) {
    throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast VARCHAR '%s' to DECIMAL(%s, %s)", stringValue, precision, scale));
}

return encodeUnscaledValue(result.unscaledValue());
```

Alternatively, we could make the messages a little more precise and do something like:

```java
BigDecimal result;
String stringValue = value.toString(UTF_8);
try {
    result = new BigDecimal(stringValue).setScale(intScale(scale), HALF_UP);
}
catch (NumberFormatException e) {
    throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast VARCHAR '%s' to DECIMAL(%s, %s). Value is not a number.", stringValue, precision, scale));
}

if (overflows(result, precision)) {
    throw new PrestoException(INVALID_CAST_ARGUMENT, format("Cannot cast VARCHAR '%s' to DECIMAL(%s, %s). Value too large.", stringValue, precision, scale));
}

return encodeUnscaledValue(result.unscaledValue());
```